cpp/src/arrow/CMakeLists.txt

```diff
@@ -712,6 +712,10 @@ if(ARROW_COMPUTE)
   add_subdirectory(compute)
 endif()
 
+if(ARROW_ENGINE)
```

Comment:
`ARROW_ENGINE` depends on `ARROW_COMPUTE`, can we set it somewhere?

Perhaps consider to place it under the compute subdirectory?

Reply:
I believe `ARROW_ENGINE` may also someday depend on datasets, ipc, parquet, and maybe even flight.  For example, a substrait plan will generally start with a scan node (datasets) and the engine may need to use spillover (ipc / parquet) and we might want to send data to or receive data from flight nodes.

Some of this we could probably avoid using more indirection (e.g. substrait consumer defines a "table provider" and the user can use the "datasets table provider" to link the two modules) but to start with it might be easier to just do the simple thing.

Either way, I think we will eventually want a standalone engine module that isn't really just a child of compute so I'm kind of in favor of it being a peer (and not a child) of compute.

See: ARROW-15238 (which this PR satisfies)

Reply:
`ARROW_ENGINE` does now depend on datasets, where a scan node is used to wrap `substrait::ReadRel`s